File: bitwise-ors-of-subarrays/bitwise-ors-of-subarrays.java
class Solution {
   public int subarrayBitwiseORs(int[] arr) {
        HashSet<Integer> res = new HashSet<>();
        HashSet<Integer> prev = new HashSet<>();
        
        for(int num: arr) {
            HashSet<Integer> cur = new HashSet<>();
            cur.add(num);
            for(int prevNum: prev) cur.add(num | prevNum);
            prev = cur;
            res.addAll(cur);
        }
        
        return res.size();
    }
}

File: 934-bitwise-ors-of-subarrays/bitwise-ors-of-subarrays.py
class Solution:
    def subarrayBitwiseORs(self, arr: List[int]) -> int:
        res = set()
        curr = set()

        for num in arr:
            # For each number, update the current set of ORs by OR-ing it with previous values
            curr = {num | x for x in curr} | {num}
            res |= curr  # Add all current ORs to the result set

        return len(res)